projects/plugins/jetpack/modules/masterbar/admin-menu/class-admin-menu.php

```diff
@@ -68,6 +68,9 @@ public function reregister_menu_items() {
 			remove_menu_page( 'link-manager.php' );
 		}
 
+		$this->sort_hidden_submenus();
+		$this->hide_unauthorized_menus();
```

Comment:
Seems this could be part of `Base_Admin_menu`?

Reply:
Imho, we should leave it as it is because, for example, in Domains only admin menu class we don't have any use for it.

Reply:
That's true at the moment, but that's also a particularity of the domain only menu class which don't register with any submenu and don't need to deal with any hidden menu.

I'm worried about future cases where other developers may need it but they won't be aware that they need to perform these calls. `Base_Admin_Menu` is intended to be the default class for the menu manipulation, and I consider this sort/hiding functionality a pretty basic behavior that should be performed always.

Reply:
Yeah, I know what you mean, I'm not too fond of it as well. My only concern with moving it to `Base_Admin_Menu` is that we would make `reregister_menu_items` a concrete again and forcing each child to call the parent, if not, the those methods will not be called. The parent method would also have to be called at the end of the child method, which other people might find it unintuitive (most of the time parent methods are called at the beginning of the method). 

Another approach would be to extract these two methods (and their implementation) into another class that would act like a Decorator or maybe Strategy class in the load.php script. 
This way, we:
- make sure that those two methods are called exactly right after the reregister method.
- We decouple this logic from admin menu classes which should contain only product specific business logic.

I think this should be pretty easy to do, we'll only have to create that class and move this add_action from the Base_Admin_Menu constructor:
```
add_action( 'admin_menu', array( $this, 'reregister_menu_items' ), 99998 )
```

Reply:
What if we move these calls to separate and independent `admin_menu` hooks that run` after `reregister_menu_items` runs??

```php
class Base_Admin_Menu {
  protected function __construct() {
    add_action( 'admin_menu', array( $this, 'reregister_menu_items' ), 99998 );
    add_action( 'admin_menu', array( $this, 'handle_hidden_submenus' ), 99999 );
    add_action( 'admin_menu', array( $this, 'hide_unauthorized_menus' ), 99999 );
  }
}
```

Reply:
It could work, I thought about it as well but I wanted to be sure they are executed immediately after the menus are manipulated otherwise they'll be in an inconsistent state until those two methods are called. Also, the order of the methods is also important because `hide_unauthorized_menus` assumes that the hidden elements are always at the end of the array; that's why it checks if the first element is hidden (meaning that we are safe to say that all elements are hidden). 

The probability of it happening is small, haha. In the end, since we only have just one place were I we use them, I thought it's just simpler to call them at the end of that method. The possibility of someone adding another action just between reregister_menu_items and our cleanup methods are small. But hey, it's still a possibility right? :D

What do you think? :)

Reply:
> the order of the methods is also important because hide_unauthorized_menus assumes that the hidden elements are always at the end of the array

If that is a required dependency, why don't we call `sort_hidden_submenus` within `hide_unauthorized_menus `?

```php
public function hide_unauthorized_menus() {
  $this->sort_hidden_submenus();
  
  // ...
}
```

> The possibility of someone adding another action just between reregister_menu_items and our cleanup methods are small

I think that's a safe risk to take. If the resulting code is more maintainable and readable, let's not deal with edge cases that are extremely unlikely to happen.

Reply:
> If that is a required dependency, why don't we call sort_hidden_submenus within hide_unauthorized_menus ?

The main purpose of the sort method is to move the hidden submenus at the end of the array, in order to be sure that hidden submenus will not be used as default menu slugs. `hide_unauthorized_menus` just benefits of that behaviour. :) 

Initially I had both functionalities in the same method, but I considered that it does too many things and I moved it to a separate method. 

Moving the sort in hide_unauthorized_menus  would make the method do too many things, imho, even though the actual implementation is in another place. I could add an anonymous function instead, so the change would like this: 
```
    add_action(
        'admin_menu', 
        function () {
            $this->sort_hidden_submenus();
            $this-> hide_unauthorized_menus();
       }, 
     99999 
   );
``` 
What do you think? 🤔

Reply:
> Moving the sort in hide_unauthorized_menus would make the method do too many things

See it this way: does `hide_unauthorized_menus` work correctly if `sort_hidden_submenus` is not called first? If not, `hide_unauthorized_menus` should call `sort_hidden_submenus`.

> I could add an anonymous function instead

Actions should be hooked using a named function, otherwise they cannot be unhooked if needed.

Reply:
> See it this way: does hide_unauthorized_menus work correctly if sort_hidden_submenus is not called first? If not, hide_unauthorized_menus should call sort_hidden_submenus.

Hmm... good point.

> Actions should be hooked using a named function, otherwise they cannot be unhooked if needed.

Oh, good to know that. :)